# classical statistical inference pipeline to a research question related to Newton's law of motion

### Research question: Does the force applied to an object affect its acceleration?

### step1: Formulate a research question or hypothesis: 
        Hypothesis: the force applied to an object is directly proportional to its acceleration

### step2: Design a study:
        a simple experiment can be designed where an object of known mass is attached to a string and pulled iwth varying amounts of force. 
        the acceleration of the object is measured using a motion sensor

### step3: collect data:
        data is collected by measuring the force applied to the object and the resulting acceleration

### step4: explore the data:
        the data is plotted on a graph to look for a linear relationship between the force and acceleration

### step5: Choose a statistical model:
        A linear regression model can be chosen to model the slope and intercept, relationship between the force and acceleration

### step6: Estimate the model parameters:
        slope and intercept, are estimated using the collected data

### step7: Test the Hypothesis:
        if the estimated slope of the linear regression model is significantly different from zero

### step8: Draw conclusions:
        hypothesis test indicates theat the estimated slope is significantly different from zero, 
        it can be conluded that there is a significant linear relationship between the force and acceleration

### step9: Communicate results:
        the findings of the study can be summarized in a research paper or report, which can be shared with others in scientific community

# A/B testing

###### 1. Testing two soit treatments to determine which produces better seed germination


###### 2. Testing two therapies to determine which suppresses cancer more effectively 

###### 3. Testing two prices to determine which yields more net profit

###### 4. Testing two web designs or headlines to determine which produces more clicks

###### 5. Testing two web ads to determine which generates more conversions

##### Subject: 
    person, plant seed, web visitor

###### Test statistic:
    binary variable
    click or no-click, buy or don't buy, fraud or no fraud,
    
    ecommerce experiment results:
        outcome  |  PriceA   | price B
        conversion  200        182
        No conversion 23,539   22406

###### Revenue/page view with price A: mean = 3.87, SD = 51.10
Revenue/page view with price B; mean = 4.11, SD = 62.98

###### multi-arm bandit algorithm

In [2]:
import numpy as np
from scipy.stats import ttest_ind

# Generate simulated data
np.random.seed(123)
original_conversion_rates = np.random.rand(1000)
redesigned_conversion_rates = np.random.normal(0.05, 0.01, size=1000)

# Conduct a two-sample t-test
t_statistic, p_value = ttest_ind(original_conversion_rates, redesigned_conversion_rates, equal_var=False)

# Print the results of the hypothesis test
if p_value < 0.05:
    print("The redesigned product page leads to a significantly higher conversion rate (p-value = {:.4f}).".format(p_value))
else:
    print("There is no significant difference in conversion rates between the original and redesigned product pages (p-value = {:.4f}).".format(p_value))


The redesigned product page leads to a significantly higher conversion rate (p-value = 0.0000).


# Hypothesis testing

### Misinterpreting randomness:
    task1: ask several friends to invent a series of 50 coin flips: write down a series of random Hs and Ts.
    
    task2: actually flip a coin 50 times and write down the results.
        
    task3: have them put the real coin flip results in one pile, and the made up results in another.
        
        
        result: it is not at all unusual to see five or six Hs or Ts in a row

### A statistical hypothesis test is further analysis of an A/B rest, or any randomize experiment.:
    to assess whether ranodm chance is a reasonable explanation for the observed difference between groups A and groups B


### Null hypothesis:involves a baseline assumption that the treatments or groups are equivalent, and any difference between the groups is due to chance. represented as H0
    1. there is no significant difference in mean test scores between male and female students
    2. the average salary of employees in a company is not affected by their years of experience
    3. there is no significant correlation between the amount of time spent exercising and the level of blood pressure in adults
    4. the propotion of defective products produced by a manufacturing process does not exceed a certain thresold
    5. the mean wight of apples grown on two different farms is not significantly different from each other

### Alternate Hypothesis: Ha
    1. the mean test scores of male and female students in a particular school are significantly different
    2. the average salary of employees in a company is affected by their years of experience
    3. there is a significant correlation between the amount of time spent exercising and the level of blood pressure in adults
    4. the proportion of defective products produced by a process exceeds a cetain thresold
    5. the mean weight of apples grown on two different farms is significantly different from each other.

### one-way test: researcher interested in whether a particular variable has a significant effect in a specific direction, + or -
    example: whether a new medication decreases the average blood pressure of patients
        H0: the medication has no effect on the blood pressure
        Ha: medication decreases the blood pressure3

### two-way test: researcher is interested in whether a particular variable has a significant effect, regardles3s of whether the effect is + or -
    example: whether a new advertising campaign increase the average sales of a product.
        H0: the advertising campaign has no effect on sales
        Ha: the advertising campaign has significant effect on sales,, either positive or negative.

# Permutation test

### example: a researcher wants to test whether there is a significant difference in the average weight between two different breeds of dogs,
    breed A 
    breed B

### H0: there is no significant difference between the average weights of the two breeds, 


### Ha: there is a significant difference

### step1: researcher collects a sample of 20 dogs from breed A and 20 dogs from breed B
    and measures their weights in pounds
    

### step2: the observed difference in means between the two breeds is 
    4.5 pounds,
    
    with breed A having an average weight of 45 pounds, and breed B having an average weight of 49.5 pounds

### step3: conducting a permuation test
    researcher randomly assigns the 40 observations to two groups of 20,
    simulating the null hypothesis of no difference between the two breeds.
    
    

### step4: researcher calculates the difference in means between two groups, and record this value. this process is repeated many times. 
    creating a distribution of differences in means under the null hypothesis

### step5: After creating the distribution of differences in means, the researcher comopares the observed difference in means of 4.5 pounds to this distribution
    
    if the observed difference is outside the range of values in the null distribution or distribution of differences in mean. 
    
    researcher can reject the null hypothesis and conclude that there is a significant difference between the weights of the two breeds

# Statistical significance and p-value

### Statistical significance is a measure of likelihood that an observed effect is not due to chance
the p-value  is used as a thresold to determine the statistical significance;

    p-value < or = specified significance level (0.05) --> stastically significant
        [ indicating that the observed effect is unlikely to have occured by chance]

### Example: we want to test whether a neew drug is effective in reducing blood pressure. 

### step1: randomly assign 100 participants to receive thenew drug or a placebo for a period of 4 weeks, and measure their blood pressure
    berfor e and after treatment.

### step2: performed two-sample t-test,
    observed test statstics:
        treatment groups: 
       and  control group: is -2.45
        and the corresponding p-value -s 0.016

### step3: p-values < 0.05 (sl)
    the observed effect is statistically significant.

# Type1 error or false positive:occures when we reject the null hypothesis when it is actually true
    significance level _ alpha = 0.05 -->5% chance of rejecting null hypthesis
    
    the probability of making type1 error is denoted by alpha and typically set at 0.05 or 0.01
    
    

# Type2 error or false negative, occurs when we fail to reject the null hypothesis when it actually false.
    the probability of making type2 error is denoted by beta and is influenced by the sample size, effect size and alpha level

### alpha --> reduce it 
--> the probability of making a type1 error but can increase the probability of making type2 error.

alpha --> increase it
--> probabiity of making type1 error increases anfd type2 decreses

alpha = 10 = 0.10


### example: hypothesis = new teaching method improves student's performance in math:
    
    H0 = there is no difference in performance between the new teaching method and the traditional method.
    Ha = there is a difference

### step1: researcher randomly selects 100 students and randomly assigns them to either the new or traditional teaching method.
    1. the mean math test score for the new teaching method group is 85 with a SD = 10
    2. the = 80 and SD =8 using old teaching method.

### step2: performs a two-sample t-test with a signficance level 
    alpha = 0.05
    

### step3: the t-test value in 
    p-value = 0.02, 
    

### step4: p-value is lesser than the alpha=0.05
    

### step5: the observed differnce in mean test scores between the two groups is statistically significant.
    reject null hypothesis 

### step6: conclusion:
        new teaching method improves student's performance in math.

### step7: possibilites of type1 and type2 errors
    type1 error = 0.05
    

### to reduce the risk of Type1 and Type2 erros:
    1. increase the sample size
    2. use a different statstical test
    3. adjust the alpha level
    

In [7]:
import numpy as np
from scipy.stats import ttest_ind

# Generate simulated data
np.random.seed(123)
new_method_scores = np.random.normal(loc=85, scale=10, size=100)
trad_method_scores = np.random.normal(loc=80, scale=8, size=100)

# Conduct a two-sample t-test
t_stat, p_val = ttest_ind(new_method_scores, trad_method_scores, equal_var=False)

# print the results
print(f"T-statistic: {t_stat:.2f}")
print(f"P-value: {p_val:.2f}")

# set significance level
alpha = 0.5

# Print the results of the hypothesis test
if p_value < alpha:
    print("The observed difference in mean test scores is statistically significant .reject the null hypothesis")
else:
    print("The observed difference in mean test scores is not statistically significant .Fail to reject the null hypothesis")

T-statistic: 3.94
P-value: 0.00
The observed difference in mean test scores is statistically significant .reject the null hypothesis


### p-values for data scientists

In [ ]:
example: a linear regression model to predict house prices based on various features such as
    dependent variable or response or target = house prices
    independent variable or predictors or features 
    1. square footage
    2. number of bedrooms
    3. location
    4. presence of swimming pool
    
case1: dataset with 100 houses records
    50 have swimming pools
    50 do not
    
    after running the regression analysis for 
    p-value = 0.02
    indicating the statistical significance level at 0.05
    
    conclusion: there is a significant association between  the presence of pool and house prices.
        swimming pool feature is more likely to be included in the model.

case2: dataset = 100 records
    10 have a pool rest do not
    p-value = 0.25 > 0.05
    conclusion: there no significant association between  the presence of pool and house prices.
        swimming pool feature is more likely to be excluded in the model.

# Multiple testing

### Alfa inflation:
    predictor variables = 20
    out come variable = 1
    
    
    when we do a series of 20 significance tests at the alpha=0.05 level, the odds are pretty fodd that 
    at least one predictor will give a Type1 error - false positive
    
    
    0.95^20 = 0.36
    
    -->the probability that atleast one predictor will (falsely) test significant is the flip isde of this proability
    = 1 - (probability that all will be nonsignificant)
    = 1- 0.36
    = 0.64
    is known as alpha inflation

### 1. Bonferroni correction: adjust the alpha for each individual test by dividing it by the number of tests being conducted

In [9]:
tests = 20
alpha = 0.05

corrected_alpha = alpha/tests

print("corrected alpha: ", corrected_alpha)

corrected alpha:  0.0025


# Degrees of freedom (df): number of values free to vary
    example: a sample with 10 values
        we know the mean of 10 values = 60, 
        there are 9 degrees of freedom
        n = 10
        df = n - 1
        df = 10 - 1
        df = 9
        
        ----> 9 df means (once you know 9 values of the sample, the 10th can be calculated and is not freee to vary.)

### we use n-1 in the denominator to estimate the sample variance :
    to make the estimate free from the bias , we use n-1 in the denominator

### Df in ANOVA:
    df for between groups = number of groups - 1
    df for within groups = total sample size - total number of groups

In [11]:
import scipy.stats as stats

# Sample data for three types of fertilizer
fertilizer1 = [4, 5, 6, 4, 5, 5]
fertilizer2 = [5, 6, 7, 6, 5, 7]
fertilizer3 = [7, 8, 6, 7, 9, 8]



### example: comparing the effectiveness of three different types of fertilizers on plant growth.
    

### step1:
    H0: there is no difference in plant growth between the three types of fertilizer.
    Ha: there is a significant difference

### step2: the mean and variance of the plant heights would be calculated for each  group (fertilizer type)

### step3: SSb - between the groups , SSW - with in the group
    SSb = sum of the squared differences between the group means and the overall mean of all plants.
    SSw = sum of the squared differences between each plant's height and the mean height of its respective group.
    SSt = sum of the squared differences between each plant's height and the overall mean height.

### step4: calculate F-stastic 
    F-statistic = SSb / SSw

### step5: determine the p-values
    p-value associated with the F-statistic would be determined using a table  or statistical software

### step6: inferences
    if the p-value < alpha (0.05)
    --> the null hypothesis would be rejected and it would be concluded that there is asignificant difference in plant growth between the three types of fertilizers

In [12]:
import scipy.stats as stats

# Sample data for three types of fertilizer
fertilizer1 = [4, 5, 6, 4, 5, 5]
fertilizer2 = [5, 6, 7, 6, 5, 7]
fertilizer3 = [7, 8, 6, 7, 9, 8]

# Calculate the ANOVA
f_value, p_value = stats.f_oneway(fertilizer1, fertilizer2, fertilizer3)

# print the results
print("F-value: ", f_value)
print("P-value: ", p_value)

F-value:  13.040540540540542
P-value:  0.0005228419958633013


In [13]:
p_value < 0.05


True

### conclusion: the null hypothesis would be rejected and it would be concluded that there is asignificant difference in plant growth between the three types of fertilizers

In [27]:
plant_data = pd.read_csv("./plants.csv")


In [28]:
plant_data

temperature humidity  growth_rate
0          low      low          2.3
1          low      low          2.4
2          low      low          2.1
3          low   medium          2.6
4          low   medium          2.5
5          low   medium          2.4
6          low     high          2.8
7          low     high          2.7
8          low     high          2.9
9       medium      low          3.1
10      medium      low          3.0
11      medium      low          2.9
12      medium   medium          3.5
13      medium   medium          3.3
14      medium   medium          3.2
15      medium     high          3.8
16      medium     high          3.9
17      medium     high          3.7
18        high      low          4.2
19        high      low          4.3
20        high      low          4.1
21        high   medium          4.6
22        high   medium          4.5
23        high   medium          4.4
24        high     high          4.9
25        high     high          4.8
26        high     high          4.7

In [ ]:
example: we want to investigate the effect of two factors, 
    1. temperature and   ---> ind var ----> low, medium, high
    2. humidity,         ---> ind vara  --> low, medium, high
    
    on the 
    growth rate of a plant species. --> dep var

In [31]:
plant_data

temperature humidity  growth_rate
0          low      low          2.3
1          low      low          2.4
2          low      low          2.1
3          low   medium          2.6
4          low   medium          2.5
5          low   medium          2.4
6          low     high          2.8
7          low     high          2.7
8          low     high          2.9
9       medium      low          3.1
10      medium      low          3.0
11      medium      low          2.9
12      medium   medium          3.5
13      medium   medium          3.3
14      medium   medium          3.2
15      medium     high          3.8
16      medium     high          3.9
17      medium     high          3.7
18        high      low          4.2
19        high      low          4.3
20        high      low          4.1
21        high   medium          4.6
22        high   medium          4.5
23        high   medium          4.4
24        high     high          4.9
25        high     high          4.8
26        high     high          4.7

In [40]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

# load the data
data = pd.read_csv("./plants.csv")

# Fit the model
model = ols('growth_rate ~ C(temperature) + C(humidity) + C(temperature):C(humidity)', data=data).fit()

# print the ANOVA table
print(sm.stats.anova_lm(model, typ=2))

                               sum_sq    df           F        PR(>F)
C(temperature)              17.708889   2.0  683.057143  1.063910e-17
C(humidity)                  1.875556   2.0   72.342857  2.484893e-09
C(temperature):C(humidity)   0.062222   4.0    1.200000  3.448641e-01
Residual                     0.233333  18.0         NaN           NaN


### Tukey's HSD test 


In [47]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Load the data
data = pd.read_csv('plants.csv')

# Fit the model
model = ols('growth_rate ~ C(temperature) + C(humidity) + C(temperature):C(humidity)', data=data).fit()

# Perform Tukey's HSD test for temperature
tukey_results_temp = sm.stats.multicomp.pairwise_tukeyhsd(data['growth_rate'], data['temperature'])
print(tukey_results_temp)

# Perform Tukey's HSD test for humidity
tukey_results_hum = sm.stats.multicomp.pairwise_tukeyhsd(data['growth_rate'], data['humidity'])
print(tukey_results_hum)


Multiple Comparison of Means - Tukey HSD, FWER=0.05
group1 group2 meandiff p-adj  lower   upper  reject
---------------------------------------------------
  high    low  -1.9778   0.0 -2.3319 -1.6237   True
  high medium  -1.1222   0.0 -1.4763 -0.7681   True
   low medium   0.8556   0.0  0.5015  1.2096   True
---------------------------------------------------
Multiple Comparison of Means - Tukey HSD, FWER=0.05
group1 group2 meandiff p-adj   lower  upper  reject
---------------------------------------------------
  high    low  -0.6444 0.2741 -1.6641 0.3752  False
  high medium  -0.3556 0.6635 -1.3752 0.6641  False
   low medium   0.2889 0.7614 -0.7308 1.3085  False
---------------------------------------------------


# Parametric vs non parametric

### non-parametric: means that it does not assume any specific distribution of the data

# Chi-square test

### example: significant association between the gender and smoking status
    H0: there is no association between gender and smoking status, 
    Ha: there is a significant association between the gender and smoking status
    

In [ ]:
step2: collect the data:
        contigency table:
                      Smokingf  |  nonsmoking | weed_ganja_cigar
            Male   |
            Female |

In [52]:
import numpy as np
from scipy.stats import chi2_contingency

# define the contingency table
observed = np.array([[100, 150, 50], [200, 300, 100]])

# perform the chi-square test
chi2, p, dof, expected = chi2_contingency(observed)


# print the results
print("Chi-square statistic: ", chi2)
print("p-value: ", p)
print("Degrees of Freedom: ", dof)
print("expected values: ", expected)

Chi-square statistic:  0.0
p-value:  1.0
Degrees of Freedom:  2
expected values:  [[100. 150.  50.]
 [200. 300. 100.]]


In [50]:
observed

array([[100, 150,  50],
       [200, 300, 100]])

### conclusion: p-values < 0.05 (chosen significance level)
    p-value = 1.0
    alpha = 0.05
    p-value is not lesser than alpha
    we failed to reject the null hypothesis and conclude that there is a evidence of a no significance association.

# Fisher's Exact test

In [55]:
import scipy.stats as stats

# create contingency table
contingency_table = [[10, 5], [8, 7]]

# perform Fisher's exact test
odds_ratio, p_value = stats.fisher_exact(contingency_table)


# print the results
print("Odds ratio: ", odds_ratio)
print("p-value: ", p_value)

Odds ratio:  1.75
p-value:  0.7103816512796233


### the odds ratio tells us that, the odds of having a cetain outcome in one gorup is 1.75 times higher than the odds of having the smae outcome in the other group.

### p-value tells us that there is no significant association between the two variable, as the p-value is greatere than commonly used singinficance level of 0.05

# Multi-Arm Bandit algorithm

### problem: there is a set of $k$ arms, each associated with a random reard distribution with an unknown mean.
    at each step $t$ , the algoritm selects an arm to play, observes a reawrd from the chose arm, and updates its 
    belief about the mean of the reward distriubtion of that arm.
    

### goal: to maximize the cumulative reward obtained over a fixed time horizon

### 1. $\epsilon$-greedy algorithm.
    it selects the arm with the highest empirical mean reward with probaibiliry $1-\epsilon$, and selects a random arm with 
    probabilithy $\epsilon$
2. UCB ( Upper confidence Bound):
    selects the arm with the highest upper confidence Bound. balances the empirical mean reward and uncertainity of the reward distribution

### Example: a company wants to determine which of two differnet ad designs to use for a new product launch.
    agend: use the MAB algorithm to allocate the traffic to each ad design,
    goal: maximizing click-through rate (CTR)

### step1: setup algorithms with two arms, or ad designs, 
    and randomly allocate traffic to each arm at the beginning, 
    As users interact with the ads, the algorithm begins to learn which ad is performing better in terms of CTR

### step2: as the algorithm collects more data, it adjusts the allocation of traffic to each addesign in real-time to maximize CTR.
    
    if one  ad desing consistently performs better than the other, the algortim will allocate more traffic to that addesign.

### result: over time, the algorithm converges to the ad design thathas the highest CTR, allowing the company to confidencly choose the best ad design for their product launch